Rewrite Contributing and readme (#4416)
opensearch-project/documentation-website: CONTRIBUTING.md

# Contributing Guidelines

Thank you for your interest in improving the OpenSource documentation! We value and appreciate all feedback and contributions from our community, including requests for additional documentation, corrections to existing content, and reports of technical issues with the documentation site. 

## Finding contributions to work on

If you’d like to contribute but don't know where to start, try browsing existing issues. Our projects use custom GitHub issue labels for status, version, type of request, and so on, but we recommend looking at any issues labeled `good first issue` first. 

## Contributing to the documentation site

There are two ways to contribute: 
1. [Create an issue](#create-an-issue) asking us to change the documentation.
2. [Contribute documentation content](#contribute-content) yourself.

> ## Create an issue

Reviewer: ```suggestion
### Create an issue
```

> Use the GitHub issue tracker to describe the change you'd like to make: 

Reviewer: ```suggestion
Use the documentation issue template to describe the change you'd like to make.
```

1. Go to https://github.com/opensearch-project/documentation-website/issues and select *New issue*.
2. Enter the requested information, including as much detail as possible, especially which version or versions the request affects.
3. Select *Submit new issue*. 

The `untriaged` label is assigned automatically. During the triage process, the documentation team will add the appropriate labels, assign the issue to a technical writer, and prioritize the request. We may follow up with you for additional information. 

> ## Contribute content

Reviewer: ```suggestion
### Contribute content
```

- [Editing files directly in GitHub](#editing-files-in-github): Best for small changes like fixing a typo or adding a parameter. This approach does not require cloning the repository and does not let you test the documentation.
- [Making changes locally and pushing to GitHub](#making-changes-locally): Best for changes you want to test locally first, like adding a table or section, or reorganizing pages. This approach requires setting up a local version of the repository and lets you test documentation.

The workflow for contributing documentation is no different than the one for contributing code:

- Make your changes
- Build locally to check your work (only possible if you are making changes locally)
- Submit a [pull request](https://github.com/opensearch-project/documentation-website/pulls) (PR)
- Maintainers review and merge your PR

If the change requires significant work, open an issue where we can first discuss your request.

### Before you start

Before contributing content, make sure to read the following resources:
- [README](README)
- [OpenSearch Project Style Guidelines](STYLE_GUIDE)
- [API Style Guide](API_STYLE_GUIDE)
- [Formatting Guide](FORMATTING_GUIDE) 

> ### Editing files in GitHub

Reviewer: I'm confused about these headings. I'll reach out offline.

Author: @hdhalter I made some changes to the structure of the doc. Please review again. Thanks.

NOTE: If you’d like to contribute but don't know where to start, try browsing existing [issues](https://github.com/opensearch-project/documentation-website/issues). Our projects use custom GitHub issue labels for status, version, type of request, and so on, but we recommend looking at any issues labeled `good first issue` first. 

## Creating an issue

Use the documentation issue template to describe the change you'd like to make.

1. Go to https://github.com/opensearch-project/documentation-website/issues and select **New issue**.
1. Enter the requested information, including as much detail as possible, especially which version or versions the request affects.
1. Select **Submit new issue**. 

The `untriaged` label is assigned automatically. During the triage process, the documentation team will add the appropriate labels, assign the issue to a technical writer, and prioritize the request. We may follow up with you for additional information. 

## Contributing content

There are two ways to contribute content depending on the magnitude of the change:

> - [Minor changes](#making-minor-changes): For small changes like fixing a typo or adding a parameter, you can edit files in GitHub directly. This approach does not require cloning the repository and does not let you test the documentation.

Reviewer: ```suggestion
- [Minor changes](#making-minor-changes): For small changes to existing files, like fixing typos or adding parameters, you can edit files in GitHub directly. This approach does not require cloning the repository and does not let you test the documentation.
```

> - [Major changes](#making-major-changes): For changes you want to test first, like reorganizing pages or adding a table or section, you can edit files locally and push the changes to GitHub. This approach requires setting up a local version of the repository and lets you test documentation.

Reviewer: ```suggestion
- [Major changes](#making-major-changes): For changes you want to test first, like adding new or reorganizing pages or adding a table or section, you can edit files locally and push the changes to GitHub. This approach requires setting up a local version of the repository and lets you test documentation.
```

- [Review process](#review-process)
- [Getting help](#getting-help)

# Contributing Guidelines

Thank you for your interest in improving the OpenSource documentation! We value and appreciate all feedback and contributions from our community, including requests for additional documentation, corrections to existing content, and reports of technical issues with the documentation site. 

You can [create an issue](#creating-an-issue) asking us to change the documentation or [contribute content](#contributing-content) yourself.

## Creating an issue

Use the documentation issue template to describe the change you'd like to make.

1. Go to https://github.com/opensearch-project/documentation-website/issues and select **New issue**.
1. Enter the requested information, including as much detail as possible, especially which version or versions the request affects.
1. Select **Submit new issue**. 

The `untriaged` label is assigned automatically. During the triage process, the documentation team will add the appropriate labels, assign the issue to a technical writer, and prioritize the request. We may follow up with you for additional information. 

> ## Finding contributions to work on

Reviewer: Can we add this as a note underneath "contributing content"?

If you’d like to contribute but don't know where to start, try browsing existing [issues](https://github.com/opensearch-project/documentation-website/issues). Our projects use custom GitHub issue labels for status, version, type of request, and so on, but we recommend looking at any issues labeled `good first issue` first. 

## Contributing content

There are two ways to contribute content depending on the magnitude of the change:

- [Minor changes](#making-minor-changes): For small changes like fixing a typo or adding a parameter, you can edit files in GitHub directly. This approach does not require cloning the repository and does not let you test the documentation.
- [Major changes](#making-major-changes): For changes you want to test first, like reorganizing pages or adding a table or section, you can edit files locally and push the changes to GitHub. This approach requires setting up a local version of the repository and lets you test documentation.

> ### Documentation workflow

Reviewer: ```suggestion
### Contribution workflow
```
?

The workflow for contributing documentation is no different than the one for contributing code:

- Make your changes

> - Build locally to check your work (only possible if you are making changes locally)

Reviewer: ```suggestion
- Build a local instance of the content to check your work (only possible if you are editing the files locally)
```

Author: Reworded.

- Submit a [pull request](https://github.com/opensearch-project/documentation-website/pulls) (PR)
- Maintainers review and merge your PR

> If the change requires significant work, open an issue where we can first discuss your request.

Reviewer: Not sure if this is needed since we already included reorganizing pages in a previous section.

### Before you start

Before contributing content, make sure to read the following resources:
- [README](README.md)
- [OpenSearch Project Style Guidelines](STYLE_GUIDE.md)
- [API Style Guide](API_STYLE_GUIDE.md)
- [Formatting Guide](FORMATTING_GUIDE.md) 

> NOTE: Please make sure that any documentation you submit is your work or work you have the rights to submit. We respect the intellectual property rights of others, and as part of contributing, we'll ask you to sign your contribution with a [Developer Certificate of Origin (DCO)](https://github.com/opensearch-project/.github/blob/main/CONTRIBUTING.md#developer-certificate-of-origin) that states you have the rights to submit this work and you understand we'll use your contribution. 

Reviewer: ```suggestion
NOTE: Please make sure that any documentation you submit is your work or work that you have the rights to submit. We respect the intellectual property rights of others, and as part of contributing, we'll ask you to sign your contribution with a [Developer Certificate of Origin (DCO)](https://github.com/opensearch-project/.github/blob/main/CONTRIBUTING.md#developer-certificate-of-origin) that states you have the rights to submit this work and you understand we'll use your contribution. 
```

### Making minor changes

> If you want to add a few paragraphs to a file, try this approach:

Reviewer: ```suggestion
If you want to make minor changes to an existing file, you can use this approach:
```

1. [Fork this repository](https://docs.github.com/en/get-started/quickstart/fork-a-repo).

1. In your fork on GitHub, navigate to the file that you want to change.

1. In the upper-right corner, select the dropdown with the pencil icon and select **Edit in place**. Edit the file.

1. In the upper-right corner, select **Commit changes...***. Enter the commit message and optional description and select **Create a new branch for this commit and start a pull request**.

### Making major changes

> If you're making major changes to the documentation and need to see the rendered HTML before submitting a pull request, you need to work in a local copy of the repository. 

Reviewer: ```suggestion
If you're adding new pages or making major changes to the documentation, we recommend that you work in a local copy of the repository and test the rendered HTML before submitting a pull request. 
```

- [Creating an issue](#creating-an-issue)
- [Contributing content](#contributing-content)

>    - [Documentation workflow](#documentation-workflow)

Reviewer: ```suggestion
   - [Contribution workflow](#contribution-workflow)
```

   - [Before you start](#before-you-start)
   - [Making minor changes](#making-minor-changes)
   - [Making major changes](#making-major-changes)
      - [Setting up your local copy of the repository](#setting-up-your-local-copy-of-the-repository)
      - [Making, viewing, and submitting changes](#making-viewing-and-submitting-changes)
- [Review process](#review-process)
   - [Style linting](#style-linting)
- [Getting help](#getting-help)

> # Contributing Guidelines

Reviewer: ```suggestion
# Contributing guidelines
```

> Thank you for your interest in improving the OpenSource documentation! We value and appreciate all feedback and contributions from our community, including requests for additional documentation, corrections to existing content, and reports of technical issues with the documentation site. 

Reviewer: ```suggestion
Thank you for your interest in improving the OpenSearch documentation! We value and appreciate all feedback and contributions from our community, including requests for additional documentation, corrections to existing content, and reports of technical issues with the documentation site. 
```

You can [create an issue](#creating-an-issue) asking us to change the documentation or [contribute content](#contributing-content) yourself.

> NOTE: If you’d like to contribute but don't know where to start, try browsing existing [issues](https://github.com/opensearch-project/documentation-website/issues). Our projects use custom GitHub issue labels for status, version, type of request, and so on, but we recommend looking at any issues labeled `good first issue` first. 

Reviewer 2: Should we mention the `help wanted` label as well?

- [Minor changes](#making-minor-changes): For small changes to existing files, like fixing typos or adding parameters, you can edit files in GitHub directly. This approach does not require cloning the repository and does not allow you to test the documentation.
- [Major changes](#making-major-changes): For changes you want to test first, like adding new or reorganizing pages or adding a table or section, you can edit files locally and push the changes to GitHub. This approach requires setting up a local version of the repository and lets you test documentation.

### Contribution workflow

The workflow for contributing documentation is no different than the one for contributing code:

- Make your changes
- Build the documentation website to check your work (only possible if you are making changes locally)
- Submit a [pull request](https://github.com/opensearch-project/documentation-website/pulls) (PR)
- Maintainers review and merge your PR

### Before you start

Before contributing content, make sure to read the following resources:
- [README](README.md)
- [OpenSearch Project Style Guidelines](STYLE_GUIDE.md)

> - [API Style Guide](API_STYLE_GUIDE.md)

Reviewer: - [Terms](TERMS.ms)

- [Formatting Guide](FORMATTING_GUIDE.md) 

> NOTE: Please make sure that any documentation you submit is your work or work that you have the rights to submit. We respect the intellectual property rights of others, and as part of contributing, we'll ask you to sign your contribution with a [Developer Certificate of Origin (DCO)](https://github.com/opensearch-project/.github/blob/main/CONTRIBUTING.md#developer-certificate-of-origin) that states you have the rights to submit this work and you understand we'll use your contribution. 

Reviewer 2: ```suggestion
NOTE: Make sure that any documentation you submit is your work or work that you have the rights to submit. We respect the intellectual property rights of others, and as part of contributing, we'll ask you to sign your contribution with a [Developer Certificate of Origin (DCO)](https://github.com/opensearch-project/.github/blob/main/CONTRIBUTING.md#developer-certificate-of-origin) that states you have the rights to submit this work and you understand we'll use your contribution. 
```

> NOTE: Make sure that any documentation you submit is your work or work that you have the rights to submit. We respect the intellectual property rights of others, and as part of contributing, we'll ask you to sign your contribution with a [Developer Certificate of Origin (DCO)](https://github.com/opensearch-project/.github/blob/main/CONTRIBUTING.md#developer-certificate-of-origin) that states you have the rights to submit this work and you understand we'll use your contribution. 

Author: This is taken from the OpenSearch main repo but I can have Weston review.

### Making minor changes

If you want to make minor changes to an existing file, you can use this approach:

1. [Fork this repository](https://docs.github.com/en/get-started/quickstart/fork-a-repo).

1. In your fork on GitHub, navigate to the file that you want to change.

> 1. In the upper-right corner, select the dropdown with the pencil icon and select **Edit in place**. Edit the file.

Reviewer 2: Why do they have to use the "Edit in place" option? They can also just select the pencil icon, which defaults to editing through GitHub automatically?

1. In the upper-right corner, select **Commit changes...***. Enter the commit message and optional description and select **Create a new branch for this commit and start a pull request**.

### Making major changes

> If you're adding new pages or making major changes to the documentation, we recommend that you work in a local copy of the repository and test the rendered HTML before submitting a pull request. 

Reviewer 2: Besides new pages and sections, what else do we define as a major change? Maybe we can revise to:

```suggestion
If you're adding new pages or making major changes to the documentation, such as adding new images, pages, sections, or styling, we recommend that you work in a local copy of the repository and test the rendered HTML before submitting a pull request. 
```

#### Setting up your local copy of the repository

Follow these steps to set up your local copy of the repository:

1. [Fork this repository](https://docs.github.com/en/get-started/quickstart/fork-a-repo) and clone your fork.

> 1. Navigate to the repository root.

Reviewer 2: ```suggestion
1. Navigate to your cloned repository.
```

Reviewer 2: Technically speaking, they could navigate to anywhere in their repository clone and still make commits.

1. Install [Ruby](https://www.ruby-lang.org/en/) if you don't already have it. We recommend [RVM](https://rvm.io/), but you can use any method you prefer:

   ```
   curl -sSL https://get.rvm.io | bash -s stable

>    rvm install 2.6

Reviewer: There is an issue to update the Ruby version to 2.7: https://github.com/opensearch-project/documentation-website/issues/4498
```suggestion
   rvm install 2.7
```

   ruby -v
   ```

1. Install [Jekyll](https://jekyllrb.com/) if you don't already have it:

   ```
   gem install bundler jekyll
   ```

> 1. Install dependencies:

Reviewer 2: ```suggestion
1. Install the Jekyll dependencies:
```

   ```
   bundle install
   ```

#### Making, viewing, and submitting changes 

Here's how to build the website, make the changes, and view them locally:

1. Build the website:

   ```
   sh build.sh    
   ```

   The build script should automatically open your web browser, but if it doesn't, open [http://localhost:4000/docs/](http://localhost:4000/docs/).

1. Create a new branch against the latest source on the main branch. 

1. Edit the Markdown files that you want to change.

> 1. When you save a file, marvel as Jekyll automatically rebuilds the site and refreshes your web browser. This process can take anywhere from 10 to 30 seconds.

Reviewer 2: The build takes longer on most machines now.

```suggestion
1. When you save a file, Jekyll automatically rebuilds the site and refreshes your web browser. This process can take anywhere from 60 - 90 seconds.
```

1. When you're happy with how everything looks, commit, [sign off](https://github.com/src-d/guide/blob/9171d013c648236c39faabcad8598be3c0cf8f56/developer-community/fix-DCO.md#how-to-prevent-missing-sign-offs-in-the-future), push your changes to your fork, and submit a pull request.

    Note that a pull request requires DCO sign-off before we can merge it. You can use the -s command line option to append this automatically to your commit message, for example `git commit -s -m 'This is my commit message'`. For more information, see https://github.com/apps/dco.

## Review process

We greatly appreciate everyone who takes the time to make a contribution. We will review all contributions as quickly as possible. If it’s a quick fix, we should be able to release the update quickly. Bigger requests might take a bit of time for us to review. 

> During the PR process, expect that there will be some back-and-forth. Please try to respond to comments in a timely fashion, and if you don't want to continue with the PR, let us know. 

Reviewer 2: ```suggestion
During the PR process, expect that there will be some back-and-forth. If you want to get your contribution merged quickly, try to respond to comments in a timely fashion, and if you don't want to continue with the PR, let us know. 
```
